packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_state_var/string_state_var.rs

```diff
@@ -158,3 +158,273 @@ impl StateVarUpdater<String, StringRequiredData> for StringStateVar {
+
+        let extending = StateVarMutableView::new_with_value("the".to_string(), false);
+        string_data.extending = Some(extending.create_new_read_only_view());
+
+        // add meta data needed for invert to work. Use arbitrary values just to test.
+        string_data._data_query_mapping_data.extending.push((3, 2));
+
+        // a call to calculate matches the extending variable
+        assert_eq!(
+            StringStateVar::calculate(&string_data),
+            StateVarCalcResult::Calculated("the".to_string())
+        );
+
+        // the essential variable is being ignored, so changes to it do not alter the calculation
+        essential.set_value("hello".to_string());
+        assert_eq!(
+            StringStateVar::calculate(&string_data),
+            StateVarCalcResult::Calculated("the".to_string())
+        );
+
+        extending.set_value("other".to_string());
+        assert_eq!(
+            StringStateVar::calculate(&string_data),
+            StateVarCalcResult::Calculated("other".to_string())
+        );
+
+        // set up a view of the state variable itself, as needed for invert
+        let state_var = StateVarMutableView::new_with_value("".to_string(), false);
+        state_var.set_requested_value("new".to_string());
+        let state_var_view = state_var.create_new_read_only_view();
+
+        // we get an Ok result from invert
+        let invert_result =
+            StringStateVar::invert(&mut string_data, &state_var_view, false).unwrap();
+
+        // the request is to change corresponds to the mapping data we made for essential
+        assert_eq!(invert_result.len(), 1);
+        assert_eq!(
+            invert_result[0],
+            DependencyValueUpdateRequest {
+                data_query_idx: 3,
+                dependency_idx: 2
+            }
+        );
+
+        // the extending value was requested to change to the new value
+        assert_eq!(
+            extending.get_requested_value().to_string(),
+            "new".to_string()
+        );
+    }
+
+    #[test]
+    fn string_state_var_extending_and_with_one_base_dependency() {
+        // set up data that consists of the essential variable, and extending, and one base variable
+        let mut string_data = StringRequiredData::default();
+        let essential = StateVarMutableView::new_with_value("".to_string(), true);
+        string_data.essential = essential.create_new_read_only_view();
+
+        let extending = StateVarMutableView::new_with_value("Hello".to_string(), false);
+        string_data.extending = Some(extending.create_new_read_only_view());
+
+        let base = StateVarMutableView::new_with_value("World".to_string(), false);
+        string_data.base = vec![base.create_new_read_only_view()];
+
+        // add meta data needed for invert to work. Use arbitrary values just to test.
+        string_data._data_query_mapping_data.extending.push((3, 2));
+        string_data._data_query_mapping_data.base.push((5, 1));
```

Comment:
It would be nice to move these setups to their own function. Any function you define inside of `mod test {}` but without `#[test]` will just be a regular function that you can call, but will not be compiled during regular builds.